components/core/src/clp/ffi/SchemaTree.hpp

```diff
@@ -91,39 +91,153 @@ class SchemaTree {
+         */
+        auto append_new_child(id_t child_id) -> void { m_children_ids.push_back(child_id); }
+
+        /**
+         * Removes the last appended child ID (if any).
+         */
+        auto remove_last_appended_child() -> void {
+            if (m_children_ids.empty()) {
+                return;
+            }
+            m_children_ids.pop_back();
+        }
+
+    private:
+        friend SchemaTree;
+
+        // Factory functions
+        /**
+         * Creates a non-root tree node.
+         * @param id
+         * @param locator
+         */
+        [[nodiscard]] static auto create(id_t id, NodeLocator const& locator) -> Node {
+            return {id, locator.get_parent_id(), locator.get_key_name(), locator.get_type()};
+        }
+
+        /**
+         * Creates a root node.
+         * @param id The ID of the root node assigned by the schema tree.
+         */
+        [[nodiscard]] static auto create_root() -> Node {
+            return {cRootId, std::nullopt, "", Type::Obj};
+        }
+
+        // Constructors
+        Node(id_t id, std::optional<id_t> parent_id, std::string_view key_name, Type type)
+                : m_id{id},
+                  m_parent_id{parent_id},
+                  m_key_name{key_name.begin(), key_name.end()},
+                  m_type{type} {}
+
+        id_t m_id;
+        std::optional<id_t> m_parent_id;
+        std::vector<id_t> m_children_ids;
+        std::string m_key_name;
+        Type m_type;
+    };
+
     /**
      * A triple---parent ID, key name, and node type---that uniquely identifies a node.
      * NOTE: We use the term "Locator" to avoid terms like "Key" or "Identifier" that are already in
      * use.
      */
     class NodeLocator {
     public:
-        NodeLocator(
-                SchemaTreeNode::id_t parent_id,
-                std::string_view key_name,
-                SchemaTreeNode::Type type
-        )
+        NodeLocator(Node::id_t parent_id, std::string_view key_name, Node::Type type)
                 : m_parent_id{parent_id},
                   m_key_name{key_name},
                   m_type{type} {}
 
-        [[nodiscard]] auto get_parent_id() const -> SchemaTreeNode::id_t { return m_parent_id; }
+        [[nodiscard]] auto get_parent_id() const -> Node::id_t { return m_parent_id; }
 
         [[nodiscard]] auto get_key_name() const -> std::string_view { return m_key_name; }
 
-        [[nodiscard]] auto get_type() const -> SchemaTreeNode::Type { return m_type; }
+        [[nodiscard]] auto get_type() const -> Node::Type { return m_type; }
 
     private:
-        SchemaTreeNode::id_t m_parent_id;
+        Node::id_t m_parent_id;
         std::string_view m_key_name;
-        SchemaTreeNode::Type m_type;
+        Node::Type m_type;
     };
 
-    // Constants
-    static constexpr SchemaTreeNode::id_t cRootId{0};
+    // Static methods
+    [[nodiscard]] constexpr static auto get_root_node_id() -> Node::id_t { return cRootId; }
```

Comment:
Why not leave `cRootId` public?

Reply:
discussed offline and realized using public `cRootId` vs. public method makes no difference.